### DISCUSSION TITLE
Formatting in VSCode

### DISCUSSION THREAD

**Discussion by aaron-hutton** [nasa/fprime#1101] [Q&A]
For anyone that develops F' projects using VSCode, have you managed to get around the `PRIVATE` #define formatting problem when using F'?

VSCode doesn't understand that `PRIVATE` used in the header files refers to the `private` keyword which then breaks the indentation and makes the file look very messy. Obviously, this is calculated at compile-time, so I don't know if it would be possible to fix, but I'd be interested in hearing if anyone has any suggestions.

**Comment by Joshua-Anderson** [comment]
No, this has been a frustrating problem! Since clang-format uses clang's parser it doesn't offer the ability to supply custom `#define` options like can be done with doxygen.

The workaround that's currently being considered is manually replacing PRIVATE before running clang-format, then swapping it back after formatting. Ex: `sed 's;PRIVATE:;private: // clang-format;' | clang-format | sed 's;private: // clang-format;PRIVATE:;'`

Long term F' may shift away from using `PRIVATE` for whitebox testing to using friend classes instead, but it's a nontrivial migration.

**Reply by aaron-hutton** [reply]
Thanks, as horrible as that is it will work for now. For anyone looking to integrate this into the actual C++ formatting (such as format on save) the steps I took were:

1. Create a file with the contents:
```
#! /bin/bash
sed -r 's;PRIVATE\s*:;private: // clang-format;' | clang-format "$@" | sed 's;private: // clang-format;PRIVATE:;'
```
I added optional whitespace before the : because a previous formatted has obviously been adding it. I'm also passing the args from VSCode to `clang-format` so changes to other VSCode settings still apply.

2. Give the file executable permissions.
3. Change the `Clang_format_path` VSCode setting to point to the file.

**Comment by ebenezerarunkumar** [comment]
Hi Everyone, 

Resurrecting an old thread here: Does the fprime team have any suggestions for how to approach formatting for an fprime project? Do the fprime developers use a `.clang-format` file? Can this be shared with the community? 

Is the approach described above used by the team to address the `PRIVATE` `#define` formatting problem?

**Reply by ebenezerarunkumar** [reply]
im especially interested in hearing if anyone has set this up with pre-commit and clang-format

**Reply by Joshua-Anderson** [reply]
https://github.com/fprime-community/fprime-tools/pull/111 has added clang-format support to fprime-util, including the special logic to avoid breaking `PRIVATE`. It's currently in fprime-tools [v3.1.2a3](https://github.com/fprime-community/fprime-tools/releases/tag/v3.1.2a3)

**Reply by ebenezerarunkumar** [reply]
thanks for that update! I tried to specify my own clang-format file, but it seems like the '--pass-through` argument does not work ...

I was trying this: `fprime-util format -f test.cpp -x --pass-through -style=file:../.clang-format`. 

The error message is: `[ERROR] No .clang-format file found in /home/visors/fprime-devel/visors-software/fprime. Override location with --pass-through --style=file:<path>.`

**Reply by LeStarch** [reply]
@thomas-bc, thoughts on this one?

**Reply by thomas-bc** [reply]
@EbenezerA99 have you tried passing through the `--style[...]` flag? Notice the double dash, where you have a single one. I have not had issues overriding the style file.

**Reply by ebenezerarunkumar** [reply]
Oops thats my bad. I tried it with the double dash and it still doesn't work - I continue to get the same error message. However, shouldn't I have originally gotten an error that `-style` was not an approved flag? 

I had a hunch that I didn't see that error because `fprime-util format` requires the .clang-format file to exist in the fprime folder(I don't currently have that because I'm working off of the master branch).

I added an empty `.clang-format` file to the fprime folder and I got a new error message:

<img width="611" alt="Screen Shot 2023-04-05 at 9 41 30 AM" src="https://user-images.githubusercontent.com/60404240/230098835-eb28add9-760d-4301-9d29-e5a6d33e5419.png">


Any thoughts?

**Reply by thomas-bc** [reply]
Huh, now that's interesting... what platform are you on and which version of clang-format are you using?
`--pass-through` is only passing through the flags to the underlying executable, it won't validate them. 
What happens when you passthrough a --style option is that the command that gets executed has two `--style` flags, one with the F' .clang-format and one with the one you're specifying. This works as expected on my machine (using only the second --style), but if it doesn't for yours, you may want to file an issue and we'll look into it.
